### PR TITLE
Module rebuild

### PR DESCRIPTION
**Pull request by DWVoid**

### CODE REVIEW

File: src/gui/neworld.cpp [hunk 1]
```diff
@@ -30,27 +29,32 @@ NEWorld::NEWorld() {
...
     getSettings();
     infostream << "Initializing...";
-    Window& window = Window::getInstance("NEWorld", 852, 480);
+    Window::getInstance("NEWorld", 852, 480);
     Texture::init();
-    void registerGUIAPI();
-    registerGUIAPI();
+	void registerGUIAPI();
+	registerGUIAPI();
+}
```
Comment (harrynull) [inline src/gui/neworld.cpp]: mixed use of tabs and spaces

File: src/Common/Filesystem.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+#endif
+
+    constexpr const char* sep =
+#if (BOOST_OS_WINDOWS)
+        "\\";
+#else
+        "/";
+#endif
+
+    std::string GetEnv(const std::string& varName) {
```
Comment (harrynull) [inline src/Common/Filesystem.cpp]: Any reason to put env-related functions into filesystem.cpp?

File: src/Common/Filesystem.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        if(auto value = std::getenv(varName.c_str()); value)
+            return value;
+        return "";
+    }
+
+    auto GetDirectoryListFromDelimitedString(const std::string& str) {
```
Comment (harrynull) [inline src/Common/Filesystem.cpp]: Inconsistent naming convention:
GetDirectoryListFromDelimitedString
search_path
makeWithString

File: src/Common/Filesystem.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+#elif (BOOST_OS_SOLARIS)
+
+#include <cstdlib>
+#include <string>
+
+filesystem::path ExecPathHelper::executable_path_worker() {
+    if (auto pathString = getexecname(); !pathString.empty())
+		return makeWithString(pathString);
```
Comment (harrynull) [inline src/Common/Filesystem.cpp]: mixed use of tabs and spaces

File: src/Common/Filesystem.h [hunk 5]
```diff
@@ -19,13 +19,22 @@
...
-namespace filesystem = boost::filesystem;
 #else
+#if __has_include(<filesystem>)
 #include <filesystem>
 namespace filesystem = std::filesystem;
+#elif __has_include(<experimental/filesystem>)
+#include <experimental/filesystem>
+namespace filesystem = std::experimental::filesystem;
+#elif __has_include(<boost/filesystem.hpp>)
+#include <boost/filesystem.hpp>
```
Comment (harrynull) [inline src/Common/Filesystem.h:35]: Note that boost/filesystem requires compilation. Avoid introducing unnecesssary dependencies especially heavy-weight libraries like boost if possible.

File: src/game/plugin/pluginmanager.cpp [hunk 6]
```diff
@@ -21,45 +21,269 @@
...
+	PluginPair(PluginPair&&) = default;
+	PluginPair& operator = (PluginPair&&) = default;
+	~PluginPair() { object.release(); lib.unload(); }
+	Library lib;
+	std::unique_ptr<PluginObject> object;
+};
+
+struct Modules {
+	std::vector<PluginPair> modules;
+	std::map<std::string, int> map;
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: Map of what? Bad naming.

File: src/game/plugin/pluginmanager.cpp [hunk 7]
```diff
@@ -21,45 +21,269 @@
...
-            strToLower(suffix);
-            if (suffix == LibSuffix) {
-                debugstream << "Loading:" << file.path().string();
-                if (loadPlugin(file.path().string()))
-                    counter++;
-            }
-        };
-    }
-    infostream << counter << " plugin(s) loaded";
+struct PluginPair {
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: PluginPair doesn't naturally make sense. Please either change to a better name or add comments to explain it.

File: src/game/plugin/pluginmanager.cpp [hunk 8]
```diff
@@ -21,45 +21,269 @@
...
-            << plugin.getData().internalName
-            << "], authored by \"" << plugin.getData().authorName << "\"";
-        return true;
+    
+    ~DefaultPluginObject() {
+        const auto finalize = mLib.get<void NWAPICALL()>("nwModuleFinalize");
+        if (finalize)
+            finalize();
+        else
+            infostream << "Module Has no unload function, skipping finalization!";
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: "Module Has no unload function" "Has" -> "has"

File: src/game/plugin/pluginmanager.cpp [hunk 9]
```diff
@@ -21,45 +21,269 @@
...
+            warningstream << "Module: " << inf.info.uri << " Failed For: " << e.what();
+            inf.stat = Status::Fail;
+        }
+        catch (...) {
+            warningstream << "Module: " << inf.info.uri << " Failed For Unknown Reason";
+            inf.stat = Status::Fail;
+        }
+    }
+
+    void verify(DependencyInfo& inf) {
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: `inf` won't be modified. Consider using `const DependencyInfo&` and adding `const` qualifier to the function.

File: src/game/plugin/pluginmanager.cpp [hunk 10]
```diff
@@ -21,45 +21,269 @@
...
+        auto& depStat = mMap[inf.uri];
+        if (depStat.stat != Status::Success) throw std::runtime_error("Dependency Load Failure");
+        if (inf.vRequired > depStat.info.thisVersion) throw std::runtime_error("Dependency Version Mismatch : Version Unsupported");
+        if (inf.vRequired < depStat.info.conflictVersion)
+            warningstream << "Dependency Version Mismatch : Confilict. Module May Behave Strangely. Use with care";
+    }
+
+    void walk() {
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: Any reason to put `walk()` as a public function? The name of  `walk` is too vague to be a public function. Consider either making it private or renaming it.

File: src/game/plugin/pluginmanager.cpp [hunk 11]
```diff
@@ -21,45 +21,269 @@
...
+                    catch (std::exception& e) {
+                        warningstream << e.what();
+                    }
+                }
+            };
+        }
+        infostream << mMap.size() << " Modules(s) Founded";
+    }
+
+    Version extractVersion(Json& json) {
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: Seems like an internal function. Consider making it private.

File: src/game/plugin/pluginmanager.cpp [hunk 12]
```diff
@@ -21,45 +21,269 @@
...
+        auto ver = getJsonValue<std::vector<int>>(json);
+        return { ver.size() >= 1 ? ver[0] : 0, ver.size() >= 2 ? ver[1] : 0,
+            ver.size() >= 3 ? ver[2] : 0, ver.size() >= 4 ? ver[3] : 0 };
+    }
+
+    PluginInfo extractInfo(const char* json) {
```
Comment (harrynull) [inline src/game/plugin/pluginmanager.cpp]: ditto

File: src/gui/gamescene.cpp [hunk 13]
```diff
@@ -225,7 +224,6 @@ GameScene::GameScene(const std::string& name, const Window& window):
         nk_labelf(ctx, NK_TEXT_LEFT, "Position: x %.1f y %.1f z %.1f",
             mPlayer.getPosition().x, mPlayer.getPosition().y, mPlayer.getPosition().z);
         nk_labelf(ctx, NK_TEXT_LEFT, "GUI Widgets: %zu", mGUIWidgets.getSize());
-        nk_labelf(ctx, NK_TEXT_LEFT, "GUI Plugins Loaded: %zu", context.plugins.getCount());
```
Comment (harrynull) [inline src/gui/gamescene.cpp]: Why is this removed?

File: src/gui/mainclient.cpp [hunk 14]
```diff
@@ -21,29 +21,53 @@
 #include <argagg.hpp>
 #include "game/context/nwcontext.hpp"
```
Comment (harrynull) [inline src/gui/mainclient.cpp:22]: inconsistent naming convention

File: src/gui/window.cpp [hunk 15]
```diff
@@ -71,12 +71,13 @@ Window::Window(const std::string& title, int width, int height)
...
     //SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
     //SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 3);
-    glewExperimental = 1;
 
     mWindow = SDL_CreateWindow(mTitle.c_str(), 100, 100, mWidth, mHeight, SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE);
     if (mWindow == nullptr)
         fatalstream << "Failed to create SDL window!" << SDL_GetError();
     Assert(mWindow != nullptr);
+	glewInit();
+	glewExperimental = 1;
```
Comment (harrynull) [inline src/gui/window.cpp]: This should be set before calling `glewInit()`.

File: src/main/entry.cpp [hunk 16]
```diff
@@ -37,48 +37,49 @@ int32_t registerBlock(const char* name, bool solid, bool translucent, bool opaqu
...
-    NWAPIEXPORT NWplugindata* NWAPICALL getInfo() {
-        auto plugin = new NWplugindata();
-        plugin->pluginName = "NEWorld";
-        plugin->authorName = "INFINIDEAS";
-        plugin->internalName = "infinideas.neworld.main";
-        plugin->pluginType = nwPluginTypeCore | nwPluginTypeGUI;
-        return plugin;
+    NWAPIEXPORT const char* NWAPICALL nwModuleGetInfo() {
+        return  R"(
+{
```
Comment (harrynull) [inline src/main/entry.cpp]: Bad aligning.

File: Externals/boost/predef.h [hunk 17]
```diff
@@ -0,0 +1,24 @@
+/*
```
Comment (harrynull) [inline Externals/boost/predef.h]: Consider changing `Externals` to `3rd_party` for better naming conventions.

**Review by harrynull — CHANGES_REQUESTED** [review]

**Comment by harrynull** [comment]
In addition, CMake compilation and CIs seem to be broken. Please fix them before merging.

File: src/Common/Filesystem.cpp (hunk 2 above)
Comment (DWVoid) [reply src/Common/Filesystem.cpp]: actually...see the fallback below

File: src/main/entry.cpp (hunk 16 above)
Comment (DWVoid) [reply src/main/entry.cpp]: Raw String.

File: src/gui/gamescene.cpp (hunk 13 above)
Comment (DWVoid) [reply src/gui/gamescene.cpp]: readded

File: src/sync_service/world/nwblock.h [hunk 18]
```diff
@@ -60,35 +60,35 @@ class BlockData {
...
-    bool isTranslucent() const { return translucent; }
+    bool isTranslucent() const { return mTranslucent; }
 
     // Opaque means it blocks light
-    bool isOpaque() const { return opaque; }
+    bool isOpaque() const { return mOpaque; }
 
     // Explode power, if it isn't a explosive set this to 0
-    int getExplodePower() const { return explodePower; }
+    int getExplodePower() const { return mExplodePower; }
```
Comment (harrynull) [inline src/sync_service/world/nwblock.h:88]: You could use `noexcept` here.

**Review by harrynull — CHANGES_REQUESTED** [review]
Don't do `3rdParty` plz orz
